src/app/core/export/export-config.controller.js

```diff
@@ -60,31 +64,48 @@ exports.exportCSV = (req, res, filename, columns, data) => {
 			s.push(null);
 		};
 
-		const sc = s.pipe(csvStream(columns));
-
-		// Pipe each row to the response
-		sc.pipe(res);
-
-		// If an error occurs, close the stream
-		s.on('error', (err) => {
-			logger.error(err, 'CSV export error occurred');
-
-			// End the download
-			res.end();
-		});
-
-		// If the client drops the connection, stop processing the stream
-		req.on('close', () => {
-			logger.info('CSV export aborted because client dropped the connection');
-			if (s != null) {
-				s.destroy();
-			}
-			// End the download.
-			res.end();
-		});
+		this.exportCSVFromStream(req, res, filename, columns, s);
 	}
 };
 
+/**
+ * Export a CSV file with rows derived from a readable stream
+ *
+ * @param {Request} req
+ * @param {Response} res
+ * @param {string} filename
+ * @param {*} columns
+ * @param {ReadableStream} stream a readable stream containing data for rows
+ */
+exports.exportCSVFromStream = (req, res, filename, columns, stream) => {
+	res.set('Content-Type', 'text/csv;charset=utf-8');
+	res.set('Content-Disposition', `attachment;filename="${filename}"`);
+	res.set('Transfer-Encoding', 'chunked');
+
+	const sc = stream.pipe(csvStream(columns));
+
+	// Pipe each row to the response
+	sc.pipe(res);
+
+	// If an error occurs, close the stream
+	stream.on('error', (err) => {
+		logger.error(err, 'CSV export error occurred');
+
+		// End the download
+		res.end();
+	});
+
+	// If the client drops the connection, stop processing the stream
+	req.on('close', () => {
+		logger.info('CSV export aborted because client dropped the connection');
+		if (stream != null) {
+			stream.destroy();
+		}
+		// End the download.
+		res.end();
+	});
+};
+
 exports.exportPlaintext = (req, res, filename, text) => {
```

Comment:
@nrwinner Can we unify these two exports into one-- or have a method to swap the stream processing implementation between csvStream/just the flat response/or whatever format we might want to add in the future? 

The only real difference between the two at current seems to be the content type and error text (which are both driven by the `req.type`, anyway), I'd rather see one block of code for the 'get stuff and create stream' and a one block for the 'output stuff' with the 'do stuff in the middle' being the only thing we need to implement based on the format requested.

Reply:
The new separated implementation is pretty neat! Thanks for cutting down on our duplicate code!